Feature dealing phase

modmo22 opened the pull request:
This closes #46.

- Created a `PokerHand` class, to represent AI hands in the poker games. Currently using the existing `UserHand` class to represent the user's hand in the poker games. We need different hands for the user and AI because they will be drawn and updated differently in some games.
- Added `TexasHoldEmManager`, this holds the logic for the completion of #46:
  -  The object `PlayerRole` will be used to track which players have which roles during a round. The small blind is always left of the dealer 
 ![image](https://github.com/user-attachments/assets/cc0dd36a-94dd-4b5a-9b0f-90c3cfb6d01c)
  - `_playerHands` is the list of the players' hands. The first hands in the list is always the user's hand. ![image](https://github.com/user-attachments/assets/be0ef31a-ab47-42bc-add8-9222eec61a11)
  - The player roles are used in the `_players` property, a dictionary with keys signifying the index of the players in `_playerHands` and values signifying which players are the dealer, and small & big blinds. ![image](https://github.com/user-attachments/assets/c5d05868-8129-4dd1-a370-272d2939e4cc)
  - Implemented `DealCards()` which deals 2 cards to each player starting with the small blind and ending with the dealer. Uses modulo to avoid an indexing error. ![image](https://github.com/user-attachments/assets/3e423bfa-b4d2-4a41-a5f1-1bf566ce5441)
- Added `Constants.cs` which includes a constant used to set the number of AI players to implement in the poker games.
- There is a lot of code refactoring, tweaks, and edits.
- I moved `BlackjackManager.cs` to the `Manager` directory.

modmo22 committed:
Set card assets to appropriate textures

modmo22 committed:
Refactored dir name to match LoadContent() calls

modmo22 committed:
Updated prologue and XML comments

modmo22 committed:
Moved BlackjackManager.cs into Manager dir

modmo22 committed:
Updated prologue and XML comments

modmo22 committed:
Corrected prologue comments

modmo22 committed:
Small refactor, added internal MainMenu object

modmo22 committed:
Small refactor

modmo22 committed:
Small edit

modmo22 committed:
Added Constants.cs

modmo22 committed:
Added PokerHand.cs

modmo22 committed:
Added TexasHoldEmManager.cs and logic for dealing cards at round start

modmo22 committed:
Small edit

modmo22 committed:
Small edit

modmo22 committed:
Inverted key-value pairs of _players dictionary

jacobwilki01 committed:
Merge pull request #60 from EECS-581-Cool-Kids/develop

Merge pull request #59 from EECS-581-Cool-Kids/master

jacobwilki01 commented on CardsCashCasino/Data/Constants.cs:
```diff
@@ -0,0 +1,6 @@
+
+public class Constants
+{
+    public const int AI_PLAYER_COUNT = 4;
```
Add a summary comment. Since the AI player count is just used for poker, we can use the comment to identify that.

jacobwilki01 commented on CardsCashCasino/Data/PokerHand.cs:
```diff
@@ -0,0 +1,25 @@
+/// <summary>
+/// The <c>PokerHand</c> class represents the AI opponents' poker hands.
+/// </summary>
+public class PokerHand : CardHand
```
I'd prefer if this name was a little more specific---instead of PokerHand, PokerAIHand or something that identifies it is just for the AI. We should be using "UserHand" for the user's hand in ALL games.

jacobwilki01 commented on CardsCashCasino/Data/UserHand.cs:
```diff
@@ -1,12 +1,12 @@
 ﻿/*
  *  Module Name: UserHand.cs
- *  Purpose: Models the user's hand of cards.
+ *  Purpose: Models the user's hand of cards. This class is used in the Blackjack game.
```
Remove this comment. UserHand should also be used in Poker.

The idea is that we only ever have one "UserHand" class that is communicated with by ALL three games. That might require some refactoring of the methods, but its intended to keep the project relatively organized.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -0,0 +1,172 @@
+        /// <summary>
+        /// The cardManager for the game.
+        /// </summary>
+        private CardManager _cardManager;
```
Let's not initialize card manager per game manager. Instead, we have a SINGLE card manager that interacts with each game manager. If we initialize per class, it makes the concept of a card manager redundant and makes communication between games hard. Inter-game communication for cards is likely not super relevant, but is for ChipManager which should also have this treatment.

modmo22 replied:
If we don't initialize a CardManager within this class, we can't access it's `Cards` object to draw cards from. If we're going to only have one CardManager, we'll need to decouple `Cards` and `DrawCard()` from it, which would defeat the purpose. See `DealCards()` for reference

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -0,0 +1,172 @@
+        /// <summary>
+        /// The chip manager for the game.
+        /// </summary>
+        private ChipManager _chipManager;
```
Like CardManager, lets not initialize this per game. Instead, initialize a single manager in the main game class. The idea is that the Managers communicate with each other through the main game class to function.

This is going to be important when we have cross-game chip inventories.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -0,0 +1,172 @@
+        /// </summary>
+        private Dictionary<PlayerRole, int> _players { get; set; }
+        
+        private List<CardHand> _playerHands { get; set; }
```
Summary comment.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -0,0 +1,172 @@
+            _ante = 0;
+
+            // Initialize the player hands. The first player is the user.
+            for (int i = 0; i < Constants.AI_PLAYER_COUNT; i++)
```
I see the intuition behind including the UserHand in this list, but I think it will make some things way harder especially as we have to make the list use the abstract class instead of the inherited class.

I am not sure the ideal solution to this, but I think we will benefit from handling the Draw / Update for UserHand at a separate step than the PokerHand classes, because they will be drawn differently. The two, rather flawed, ideas I had were to create a "nextHand" pointer in either OR keep track of when the user plays in the order. I am sure there is a much better way to solve that problem, but it is probably worth some discussion in the Discord.

modmo22 replied:
I figured that since both UserHand and PokerHand are derived from the abstract class, so we should be able to be acted on in the same way that the PokerHand class would, right? Can you explain more about what the complication would be? 

And maybe I'm not thinking about this correctly, but would we have to implement the Draw /Update methods in different places between two hands as long as the logic is different? I'm thinking of something that was similar to how we drew the ships according to whether or not they were sunk or whose ships they belonged to, except with if all players are all in, what part of the round it is, etc.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -0,0 +1,172 @@
+            // Initialize the player hands. The first player is the user.
+            for (int i = 0; i < Constants.AI_PLAYER_COUNT; i++)
+            {
+                if (i == 0)
```
If we stick with this implementation, this won't work. C# lists index first, add second, so it will error out when you try to do this. We are also creating a second user hand object that is distinct from the one created above. That will create all sorts of problems.

The easy solution to this is to just add the UserHand to the list first, then iterate over the count of AI_PLAYER_COUNT and do _playerHands.Add(new PokerHand());

modmo22 replied:
Good catch with the creation of a second user hand.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -0,0 +1,172 @@
+            };
+        }
+
+        public void StartNewGame()
```
Few thoughts about this method:
1. Summary comment!
2. This seems incomplete OR we need to remove the white space at the bottom. Shouldn't there be something calling the first move of the game?
3. Clearing _playerHands and not regenerating it seems concerning.

modmo22 replied:
Wow. I completely neglected to finish this method. I'll make a wrapper function for creating the _playerHands list. and place it here

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -0,0 +1,172 @@
+                _playerHands[playerIndex].AddCard(_cardManager.DrawCard());
+            }
+        }
+        
```
remove whitespace.

jacobwilki01 committed:
beginning

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -0,0 +1,172 @@
+            }
+            
+            // Initialize the player roles. The user starts as the dealer. 
+            _players = new Dictionary<PlayerRole, int>
```
Really, really not a fan of this implementation. Player role's mapped to ints makes it unnecessarily confusing. 

Instead, we can just have "_currentDealer", "_currentSmallBlind", and "_currentBigBlind" integers stored outside of the enum / dictionary mapping. 

That saves some complexity, and means we can also pretty easily update the values when a round is over by just using "++" and a simple check to make sure we don't go over the size of the list. We already have to do this "++" anyways, so why the dictionary?

This would require renaming some of the above variables to be more descriptive: "_currentSmallBlindBet" and "_currentSmallBlindPlayer" would solve.

modmo22 replied:
Simpler implementation and would allow a much easier randomization of which player starts out as the dealer upon initialization of Texas Hold Em. Good call.

jacobwilki01 commented on CardsCashCasino/Manager/TexasHoldEmManager.cs:
```diff
@@ -0,0 +1,172 @@
+                int playerIndex = (currentIndex + i) % _players.Count;
+                
+                // Deals two cards to each player.
+                _playerHands[playerIndex].AddCard(_cardManager.DrawCard());
```
This is incorrect implementation. you don't deal both cards at the same time, you give each player one card and then go around a second time and give each player a second card.

Mostly a technicality because the speed of computers means this is instant, but I think we are likely going to want to implement some sort of pause in the future to make the game look better when updating the screen, so we are going to want to have the implementation correct.

jacobwilki01 committed:
More refactoring

modmo22 committed:
Replaced dictionary of player roles with class properties, logic refactoring, renamed PokerHand.cs

jacobwilki01 committed:
A lot of it.

modmo22 committed:
Changed the assetNames in LoadContent back to the temp texture

modmo22 committed:
Deleted unused code

jacobwilki01 committed:
Big push of changes

jacobwilki01 committed:
Broken split functionality.

Then need to update locations of things dynamically, shouldn't be too bad.

Then need to figure out end of round transition.

jacobwilki01 committed:
Dynamic window sizing

jacobwilki01 committed:
Fix splitting issues

jacobwilki01 committed:
add a lot of constants

modmo22 committed:
Merge pull request #61 from EECS-581-Cool-Kids/blackjack-refactor

Blackjack refactor

modmo22 committed:
Merged master into feature branch

modmo22 committed:
Refactored prior TexasHoldEmManager logic, added constants, added logic for THE in CCC

modmo22 committed:
Continued refactor

jacobwilki01 commented on CardsCashCasino/CardCashCasinoGame.cs:
```diff
@@ -73,6 +112,14 @@ protected override void Update(GameTime gameTime)
+            // same for the main menu
+            if (_blackjackManager.IsPlaying)
+                _blackjackManager.Update();
+            // same for texas hold em
```
you can remove this comment about texas hold em

jacobwilki01 commented on CardsCashCasino/CardCashCasinoGame.cs:
```diff
@@ -81,7 +128,15 @@ protected override void Update(GameTime gameTime)
+            // same for the main menu
+            if (_blackjackManager.IsPlaying)
+                _blackjackManager.Draw(_spriteBatch!);
+            // same for texas hold em
```
have we fleshed out the draw functionality for texas hold em yet?

modmo22 replied:
Yes. It's in this PR,, but I can't get anything to show to the screen. More than likely just missing a ton of logic for it.

modmo22 replied:
Oh. I'm an idiot. That's why it isn't drawing. I'll add that line in now.

jacobwilki01 reviewed (approved):
Make sure you add summary comments to everything.

modmo22 committed:
Update CardCashCasinoGame.cs

Removed unnecessary comment

modmo22 committed:
Update CardCashCasinoGame.cs

Updated Draw method
